### DISCUSSION TITLE
CPD - Supporting other extensions (like .mjs and .cjs) with javascript

### DISCUSSION THREAD

**Discussion by stephen-carter-at-sf** [pmd/pmd#5297] [Q&A]
Looking at the documentation https://docs.pmd-code.org/latest/pmd_userdocs_cpd.html I don't see any information with regards to file extensions... just languages. So if javascript is a supported language, does it support other file extensions beside just '.js'?

I notice the cpd-gui only shows "js" as an extension when I select JavaScript.

So do other javascript files like .mjs and .cjs files get picked up?

I see https://github.com/pmd/pmd/blob/73dbc7a67fc37b852e7a63f040117315b3d184cf/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/EcmascriptLanguageModule.java#L22 only has js extension listed... whereas https://github.com/pmd/pmd/blob/master/pmd-html/src/main/java/net/sourceforge/pmd/lang/html/HtmlLanguageModule.java has many extensions. So maybe we just need to update the EcmascriptLanguageModule with additional extensions?

**Comment by adangel** [comment]
Yes, we'll need to provide more extensions for the EcmascriptLanguageModule. For Javascript/Ecmascript, I think, this is ok, as the extensions `.mjs` and `.cjs` are common enough and js specific. PRs are welcomed.

Note, we have a problem though for other languages like `xml` - specific xml-based languages use usually a custom extension and won't be picked up for that reason. But adding all the possible extensions for xml-based languages probably doesn't make sense - hence we add the option to override the extension based selection via [`--force-language`](https://docs.pmd-code.org/latest/pmd_userdocs_cli_reference.html#-force-language). Refs #4945, #461, #1540
